components/places/examples/places_utils.rs

```diff
@@ -0,0 +1,232 @@
+            title: dm.title,
+            children: dm
+                .children
+                .into_iter()
+                .filter_map(|c| convert_node(c))
+                .collect(),
+        }),
+    })
+}
+
+fn do_import(db: &PlacesDb, root: BookmarkTreeNode) -> Result<()> {
+    // We need to import each of the sub-trees individually.
+    // Later we will want to get smarter around guids - currently we will
+    // fail to do this twice due to guid dupes - but that's OK for now.
+    let folder = match root {
+        BookmarkTreeNode::Folder(folder_node) => folder_node,
+        _ => {
+            println!("Imported node isn't a folder structure");
+            return Ok(());
+        }
+    };
+    let is_root = match folder.guid {
+        Some(ref guid) => BookmarkRootGuid::Root == *guid,
+        None => false,
+    };
+    if !is_root {
+        // later we could try and import a sub-tree.
+        println!("Imported tree isn't the root node");
+        return Ok(());
+    }
+
+    for sub_root_node in folder.children {
+        let sub_root_folder = match sub_root_node {
+            BookmarkTreeNode::Folder(folder_node) => folder_node,
+            _ => {
+                println!("Child of the root isn't a folder - skipping...");
+                continue;
+            }
+        };
+        println!("importing {:?}", sub_root_folder.guid);
+        insert_tree(db, &sub_root_folder)?
+    }
+    Ok(())
+}
+
+fn run_desktop_import(db: &PlacesDb, matches: &clap::ArgMatches) -> Result<()> {
+    let filename = matches.value_of("input-file").unwrap();
+    println!("import from {}", filename);
+
+    let file = File::open(filename)?;
+    let reader = BufReader::new(file);
+    let m: DesktopItem = serde_json::from_reader(reader)?;
+    // convert mapping into our tree.
+    let root = match convert_node(m) {
+        Some(node) => node,
+        None => {
+            println!("Failed to read a tree from this file");
+            return Ok(());
+        }
+    };
+    do_import(db, root)
+}
+
+fn run_native_import(db: &PlacesDb, matches: &clap::ArgMatches) -> Result<()> {
+    let filename = matches.value_of("input-file").unwrap();
+    println!("import from {}", filename);
+
+    let file = File::open(filename)?;
+    let reader = BufReader::new(file);
+
+    let root: BookmarkTreeNode = serde_json::from_reader(reader)?;
+    do_import(db, root)
+}
+
+fn run_native_export(db: &PlacesDb, matches: &clap::ArgMatches) -> Result<()> {
+    let filename = matches.value_of("output-file").unwrap();
+    println!("export to {}", filename);
+
+    let file = File::create(filename)?;
+    let writer = BufWriter::new(file);
+
+    let tree = fetch_tree(db.conn(), &BookmarkRootGuid::Root.into())?.unwrap();
+    serde_json::to_writer_pretty(writer, &tree)?;
+    Ok(())
+}
+
+fn main() -> Result<()> {
+    init_logging();
+
+    let matches = clap::App::new("places_utils")
```

Comment:
In the future we should start using structopt (which generates the `clap` calls), as in https://github.com/mozilla/application-services/blob/master/testing/sync-test/src/main.rs#L54-L90. But this is fine unless you feel inspired to change it, just a note.